### PR TITLE
Add show tables access control

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

**Comment by kokosing** [comment]
#2899 was merged prematurely. Sorry about that.

In https://github.com/prestosql/presto/pull/2899#issuecomment-589683939 I said that every user should be allowed to run query against `information_schema` tables which is correct. But running `SHOW TABLES` for schema that user is not allowed to see makes an impression that schema is accessible, but there is no visible tables to user.

Notice that, typically schema is also visible to user when there is any table that is visible within that schema. Such check is quite complicated and so it is not implemented for most access controls (like file based and even sql-standard in hive).

**Comment by dain** [comment]
I would rename this to something like "Add show tables access control".  I don't see is as restoring anything as old call in #2899 was not what we wanted or needed.  If we do want to restore this, I'd do a straight revert instead.

**Comment by kokosing** [comment]
Updated

File: presto-main/src/main/java/io/prestosql/sql/rewrite/ShowQueriesRewrite.java [hunk 1]
```diff
@@ -226,6 +230,10 @@ protected Node visitShowGrants(ShowGrants showGrants, Void context)
 
                 catalogName = qualifiedTableName.getCatalogName();
 
+                accessControl.checkCanShowTables(
```
Comment (dain) [inline presto-main/src/main/java/io/prestosql/sql/rewrite/ShowQueriesRewrite.java]: Add a comment here and below that this check is wrong, and we should add a new method `checkCanShowGrants`

File: presto-spi/src/main/java/io/prestosql/spi/security/SystemAccessControl.java [hunk 2]
```diff
@@ -246,6 +246,20 @@ default void checkCanSetTableComment(SystemSecurityContext context, CatalogSchem
...
+     * <p>
+     * NOTE: This method is only present to give users an error message when listing is not allowed.
+     * The {@link #filterTables} method must filter all results for unauthorized users,
+     * since there are multiple ways to list tables.
+     *
+     * @throws AccessDeniedException if not allowed
+     */
+    default void checkCanShowTables(SystemSecurityContext context, CatalogSchemaName schema)
+    {
+        AccessDeniedException.denyShowTables(schema.toString());
```
Comment (dain) [inline presto-spi/src/main/java/io/prestosql/spi/security/SystemAccessControl.java]: Static import

File: presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java [hunk 3]
```diff
@@ -181,6 +181,16 @@ public static void denyCommentTable(String tableName, String extraInfo)
...
     }
 
+    public static void denyShowTables(String schemaName)
+    {
+        denyShowTables(schemaName, null);
+    }
+
+    public static void denyShowTables(String schemaName, String extraInfo)
+    {
+        throw new AccessDeniedException(format("Cannot show metadata of tables in %s%s", schemaName, formatExtraInfo(extraInfo)));
```
Comment (dain) [inline presto-spi/src/main/java/io/prestosql/spi/security/AccessDeniedException.java]: `Cannot show tables in...`

**Pull request merged**